fastcrypto/src/vdf/wesolowski.rs

```diff
@@ -107,13 +114,53 @@ impl WesolowskiVDF<QuadraticForm> {
     }
 }
 
-/// Compute the prime modulus used in proving and verification. This is a Fiat-Shamir construction
-/// to make the Wesolowski VDF non-interactive.
-fn get_challenge(x: &[u8], y: &[u8], bit_length: usize) -> BigInt {
-    let mut seed = vec![];
-    seed.extend_from_slice(x);
-    seed.extend_from_slice(y);
-    hash_prime(&seed, bit_length, &[bit_length - 1]).expect("The length should be a multiple of 8")
+pub trait FiatShamir<G: ParameterizedGroupElement + UnknownOrderGroupElement> {
+    /// Compute the prime modulus used in proving and verification. This is a Fiat-Shamir construction
+    /// to make the Wesolowski VDF non-interactive.
+    fn compute_challenge<F>(vdf: &WesolowskiVDF<G, F>, input: &G, output: &G) -> G::ScalarType;
+}
+
+/// Size of the random prime modulus B used in proving and verification.
+const B_BITS: usize = 264;
+
+/// Implementation of the Fiat-Shamir challenge generation compatible with chiavdf.
+/// Note that this implementation is weak, meaning that not all public parameters are used in the
+/// challenge generation. This is not secure if an adversary can influence the public parameters.
+/// See https://eprint.iacr.org/2023/691.
+pub struct ChiaFiatShamir;
+
+impl FiatShamir<QuadraticForm> for ChiaFiatShamir {
+    fn compute_challenge<F>(
+        _vdf: &WesolowskiVDF<QuadraticForm, F>,
+        input: &QuadraticForm,
+        output: &QuadraticForm,
+    ) -> BigInt {
+        let mut seed = vec![];
+        seed.extend_from_slice(&input.as_bytes());
+        seed.extend_from_slice(&output.as_bytes());
+        hash_prime(&seed, B_BITS, &[B_BITS - 1]).expect("The length should be a multiple of 8")
+    }
+}
+
+/// Implementation of the Fiat-Shamir challenge generation for usage with Wesolowski's VDF construction.
+/// The implementation is strong, meaning that all public parameters are used in the challenge generation.
+/// See https://eprint.iacr.org/2023/691.
+pub struct StrongFiatShamir<const CHALLENGE_SIZE: usize>;
+
+impl<const CHALLENGE_SIZE: usize> FiatShamir<QuadraticForm> for StrongFiatShamir<CHALLENGE_SIZE> {
+    fn compute_challenge<F>(
+        vdf: &WesolowskiVDF<QuadraticForm, F>,
+        input: &QuadraticForm,
+        output: &QuadraticForm,
+    ) -> BigInt {
+        let mut seed = vec![];
+        seed.extend_from_slice(&input.as_bytes());
```

Comment:
let's make sure that the input to the hash has 1:1 encoding, either by prefixing vectors with their length (fixed u32) or by fixed size inputs (per discr.)

Reply:
Actually, the length is fixed size per discriminant size. The input and output has length as computed in `serialized_length` which depends only (and is non-decreasing) on the discriminant length, and the number of iterations is [always 8 bytes](https://doc.rust-lang.org/std/primitive.u64.html#method.to_be_bytes).

Reply:
sg, let's document this in the function